File: 106-build-tree/Solution.java
//官方解答
import java.util.HashMap;

class TreeNode {
    int val;
    TreeNode left;
    TreeNode right;
    TreeNode(int val){
        this.val = val;
    }
    TreeNode(int val, TreeNode left, TreeNode right){
        this.val = val;
        this.left = left;
        this.right = right;
    }
}

class Solution106 {
    public TreeNode buildTree(int[] inorder, int[] postorder) {
        var map = new HashMap<Integer, Integer>();
        for (int i = 0; i < inorder.length; i++) {
            map.put(inorder[i], i);
        }
        System.out.println(map);
        return buildTree(postorder, 0, postorder.length - 1, map, 0, inorder.length - 1);
    }
    private TreeNode buildTree(int[] postorder, int postLeft, int postRight,
                               HashMap<Integer, Integer> map, int inLeft, int inRight)
    {
        if (postLeft > postRight || inLeft > inRight)
            return null;
        int rootVal = postorder[postRight];
        var root = new TreeNode(rootVal);
        int pIndex = map.get(rootVal);
        System.out.printf("rootVal = %d, pIndex = %d\n", rootVal, pIndex);
        root.left = buildTree(postorder, postLeft, postLeft + pIndex - inLeft - 1, map, inLeft, pIndex-1);
        root.right = buildTree(postorder, postLeft + pIndex - inLeft, postRight - 1, map, pIndex+1, inRight);
        return root;
    }
    static void main(String[] args) {
        int[] postOrder = {9, 15, 7, 20, 3};
        int[] inOrder = {9, 3, 15, 20, 7};
        Solution106 sol = new Solution106();
        sol.buildTree(inOrder, postOrder);
    }
}
/*
preorder:
左子樹節點數 = pIndex - inLeft
左子樹範圍:
[postLeft, postLeft + 左子樹節點數 - 1]
右子樹範圍:
[postLeft + 左子樹節點數, postRight - 1]
*/

File: 106-build-tree/solution.go
package main

import (
	"fmt"
	"gopractice/datastructure/Tree"
)

type TreeNode = Tree.TreeNode

func buildTree(inorder []int, postorder []int) *TreeNode {
	m := make(map[int]int)
	for i, v := range inorder {
		m[v] = i
	}
	fmt.Println(m)
	return recursion(postorder, 0, len(postorder)-1, m, 0, len(inorder)-1)
}
func recursion(postorder []int, postLeft int, postRight int, m map[int]int, inLeft int, inRight int) *TreeNode {
	if postLeft > postRight || inLeft > inRight {
		return nil
	}
	rootVal := postorder[postRight]
	root := &TreeNode{Val: rootVal}
	pIndex := m[rootVal]
	root.Left = recursion(postorder, postLeft, pIndex-1-inLeft+postLeft, m, inLeft, pIndex-1)
	root.Right = recursion(postorder, pIndex-inLeft+postLeft, postRight-1, m, pIndex+1, inRight)
	return root
}
func main() {
	inorder := []int{9, 3, 15, 20, 7}
	postorder := []int{9, 15, 7, 20, 3}
	buildTree(inorder, postorder)
}

/*
inorder:
[inLeft, pIndex-1][pIndex][pIndex+1, inRight]
postorder:
[postLeft, pIndex - 1 - inLeft + postLeft][pIndex -  inLeft + postLeft, postRight - 1][postRight]

105, 106題, 寫了很多遍但是從來沒有在真正意義上理解

[遞歸樹]
buildTree([9,3,15,20,7], [9,15,7,20,3])
│
└─ recursion(post=[9,15,7,20,3], postL=0, postR=4, inL=0, inR=4)
   root=3 (postorder[4])
   pIndex=1 (inorder中3的位置)
   │
   ├─ Left: recursion(post=[9,15,7,20,3], postL=0, postR=0, inL=0, inR=0)
   │   root=9 (postorder[0])
   │   pIndex=0
   │   ├─ Left: recursion(... postL=0, postR=-1) → nil
   │   └─ Right: recursion(... postL=0, postR=-1) → nil
   │
   └─ Right: recursion(post=[9,15,7,20,3], postL=1, postR=3, inL=2, inR=4)
       root=20 (postorder[3])
       pIndex=3
       │
       ├─ Left: recursion(post=[9,15,7,20,3], postL=1, postR=1, inL=2, inR=2)
       │   root=15 (postorder[1])
       │   pIndex=2
       │   ├─ Left: nil
       │   └─ Right: nil
       │
       └─ Right: recursion(post=[9,15,7,20,3], postL=2, postR=2, inL=4, inR=4)
           root=7 (postorder[2])
           pIndex=4
           ├─ Left: nil
           └─ Right: nil
*/
